File: examples/langgraph_openshell/adapter/runtime.py
# SPDX-FileCopyrightText: Copyright (c) 2026, NVIDIA CORPORATION & AFFILIATES. All rights reserved.
# SPDX-License-Identifier: Apache-2.0

"""Persistent NeMo Fabric lifecycle for the Portable Courier graph."""

from __future__ import annotations

import json
from pathlib import Path
from typing import Any

from langgraph.graph.state import CompiledStateGraph
from nemo_fabric_adapter_contract.models import AgentArtifact
from nemo_fabric_adapter_contract.models import AgentConfig
from nemo_fabric_adapter_contract.models import AgentRunRequest
from nemo_fabric_adapter_contract.models import AgentRunResult
from nemo_fabric_adapter_contract.models import AgentRunStatus
from nemo_fabric_adapter_contract.models import RuntimeContext
from nemo_fabric_adapters.common import lifecycle

from examples.langgraph_openshell.agent.graph import build_courier_graph


def main() -> None:
    """Serve one persistent graph runtime."""

    lifecycle.serve(PortableCourierRuntime, config_loader=AgentConfig.from_mapping)


class PortableCourierRuntime:
    """One compiled graph and checkpoint namespace per Fabric runtime."""

    def __init__(self) -> None:
        self._runtime_id: str | None = None
        self._artifact_root: Path | None = None
        self._graph: CompiledStateGraph | None = None

    async def start(self, payload: dict[str, Any]) -> None:
        config = payload.get("config")
        if not isinstance(config, AgentConfig):
            raise lifecycle.LifecycleError(
                "portable_courier_invalid_config",
                "Portable Courier requires a validated AgentConfig",
            )
        try:
            context = RuntimeContext.from_mapping(payload.get("runtime_context"))
        except Exception as error:
            raise lifecycle.LifecycleError(
                "portable_courier_invalid_runtime_context",
                "Portable Courier requires a valid RuntimeContext",
            ) from error
        if context.environment.artifacts is None:
            raise lifecycle.LifecycleError(
                "portable_courier_artifact_root_required",
                "Portable Courier requires an environment artifact root",
            )
        self._runtime_id = context.runtime_id
        self._artifact_root = Path(context.environment.artifacts)
        self._graph = build_courier_graph()

    async def invoke(
        self,
        request: AgentRunRequest,
        context: RuntimeContext,
    ) -> AgentRunResult:
        if (
            self._graph is None
            or self._runtime_id is None
            or self._artifact_root is None
        ):
            raise lifecycle.LifecycleError(
                "portable_courier_runtime_not_started",
                "Portable Courier is not started",
            )
        if context.runtime_id != self._runtime_id:
            raise lifecycle.LifecycleError(
                "portable_courier_runtime_mismatch",
                "The invocation does not match the active Portable Courier runtime",
            )
        if not isinstance(request.input, str) or request.input not in {
            "route",
            "deliver",
        }:
            raise lifecycle.LifecycleError(
                "portable_courier_invalid_request",
                "Portable Courier accepts only `route` or `deliver`",
            )

        result = self._graph.invoke(
            {"command": request.input, "attempts": []},
            config={"configurable": {"thread_id": self._runtime_id}},
        )
        output = {
            "command": request.input,
            "attempts": result.get("attempts", []),
            "selected_route": result.get("selected_route"),
            "delivery_status": result.get("delivery_status", "routed"),
        }
        artifacts: list[AgentArtifact] = []
        if request.input == "deliver" and result.get("delivery_status") == "delivered":
            self._artifact_root.mkdir(parents=True, exist_ok=True)
            receipt = {
                "runtime_id": self._runtime_id,
                "delivery_status": "delivered",
                "selected_route": result["selected_route"],
                "attempts": result.get("attempts", []),
            }
            (self._artifact_root / "delivery-receipt.json").write_text(
                json.dumps(receipt, indent=2) + "\n",
                encoding="utf-8",
            )
            artifacts.append(
                AgentArtifact(
                    name="delivery-receipt",
                    kind="receipt",
                    path="delivery-receipt.json",
                    media_type="application/json",
                )
            )
        return AgentRunResult(
            status=AgentRunStatus.SUCCEEDED,
            output=output,
            artifacts=artifacts,
        )

    async def stop(self) -> None:
        self._runtime_id = None
        self._artifact_root = None
        self._graph = None


if __name__ == "__main__":
    main()

File: adapters/typescript/kilo/src/runtime.ts
// SPDX-FileCopyrightText: Copyright (c) 2026, NVIDIA CORPORATION & AFFILIATES. All rights reserved.
// SPDX-License-Identifier: Apache-2.0

import type { AgentRunRequest, AgentRunResult, AgentUsage, RuntimeContext } from "nemo-fabric-adapter-contract";
import { LifecycleError, type AdapterRuntime, type AdapterStartInput } from "nemo-fabric-adapters-common";

export interface KiloPromptOutcome { text?: string; error?: boolean; usage?: AgentUsage }
export interface KiloSessionHandle { id: string; prompt(text: string): Promise<KiloPromptOutcome>; stop(): Promise<void> }
export interface KiloSessionFactory { create(input: AdapterStartInput): Promise<KiloSessionHandle> }

function failed(code: string, message: string, usage?: AgentUsage): AgentRunResult {
  return {status: "failed", output: null, error: {code, message, retryable: false}, ...(usage === undefined ? {} : {usage})};
}

export class KiloAdapterRuntime implements AdapterRuntime {
  private readonly factory: KiloSessionFactory;
  private session?: KiloSessionHandle;

  constructor(factory: KiloSessionFactory) { this.factory = factory; }

  async start(input: AdapterStartInput): Promise<void> {
    if (this.session !== undefined) throw new LifecycleError("kilo_already_started", "Kilo Code adapter runtime is already started");
    this.session = await this.factory.create(input);
  }

  async invoke(request: AgentRunRequest, _context: RuntimeContext): Promise<AgentRunResult> {
    if (this.session === undefined) throw new LifecycleError("kilo_not_started", "Kilo Code adapter runtime is not started");
    if (typeof request.input !== "string") return failed("kilo_unsupported_input", "The Kilo Code adapter accepts only plain-text input");
    const session = this.session;
    let outcome: KiloPromptOutcome;
    try {
      outcome = await session.prompt(request.input);
    } catch (error) {
      this.session = undefined;
      try { await session.stop(); } catch { /* Preserve the prompt failure. */ }
      if (error instanceof LifecycleError) throw error;
      throw new LifecycleError("kilo_session_failed", "Kilo Code session communication failed", {retryable: true});
    }
    if (outcome.error === true) return failed("kilo_model_error", "Kilo Code model invocation failed", outcome.usage);
    if (outcome.text === undefined || outcome.text.length === 0) {
      return failed("kilo_no_assistant_response", "Kilo Code completed without a final assistant text response", outcome.usage);
    }
    return {status: "succeeded", output: {response: outcome.text}, ...(outcome.usage === undefined ? {} : {usage: outcome.usage})};
  }

  async stop(): Promise<void> {
    const session = this.session;
    this.session = undefined;
    if (session !== undefined) await session.stop();
  }
}
